crates/solver/src/liquidity/zeroex.rs

```diff
@@ -47,17 +50,64 @@ impl ZeroExLiquidity {
             .filter_map(|order| TokenPair::new(order.buy_token, order.sell_token));
         let relevant_pairs = self.base_tokens.relevant_pairs(user_order_pairs);
 
-        let filtered_zeroex_orders = zeroex_orders
+        let order_buckets = self.generate_order_buckets(zeroex_orders, relevant_pairs);
+        let filtered_zeroex_orders = self.get_useful_orders(order_buckets, 5);
+
+        Ok(filtered_zeroex_orders)
+    }
+
+    fn generate_order_buckets(
+        &self,
+        zeroex_orders: impl Iterator<Item = OrderRecord>,
+        relevant_pairs: HashSet<TokenPair>,
+    ) -> OrderBuckets {
+        // divide orders in buckets
+        let mut buckets: OrderBuckets = Default::default();
+        zeroex_orders
             .filter(|record| {
                 match TokenPair::new(record.order.taker_token, record.order.maker_token) {
                     Some(pair) => relevant_pairs.contains(&pair),
                     None => false,
                 }
             })
-            .filter_map(|record| self.record_into_liquidity(record))
-            .collect();
+            .for_each(|order| {
+                let bucket = buckets
+                    .entry((order.order.maker_token, order.order.taker_token))
+                    .or_insert(vec![]);
+                bucket.push(order);
+            });
+        buckets
+    }
 
-        Ok(filtered_zeroex_orders)
+    /// Get the `orders_per_type` best priced and biggest volume orders.
+    fn get_useful_orders(
+        &self,
+        order_buckets: OrderBuckets,
+        orders_per_type: usize,
+    ) -> Vec<Liquidity> {
+        let mut filtered_zeroex_orders = vec![];
+        order_buckets.into_values().for_each(|mut orders| {
```

Comment:
I slightly prefer a normal `for ...` loop rather than `for_each`. They are mostly equivalent (return becomes continue) but it feels more correct with a real for loop unless there is a good reason for the other way.